pkg/cmd/cli/secrets/source.go

```diff
@@ -0,0 +1,230 @@
+	cmd.Flags().StringVar(&o.Password, "password", "", "password or token for Git authentication")
+	cmd.Flags().StringVar(&o.PrivateKeyPath, "ssh-privatekey", "", "path to the ssh private key")
+	cmd.Flags().StringVar(&o.CertificatePath, "ca-cert", "", "path to the certificate file")
+	cmd.Flags().StringVar(&o.GitConfigPath, "gitconfig", "", "path to the .gitconfig file")
+	cmd.Flags().BoolVarP(&o.PromptForPassword, "prompt", "", false, "prompt for password or token")
+	kcmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *CreateSourceSecretOptions) CreateSourceSecret() error {
+	secret, err := o.MakeSourceSecret()
+	if err != nil {
+		return err
+	}
+
+	if _, err := o.SecretsInterface.Create(secret); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(o.GetOut(), "secret/%s\n", secret.Name)
+	return nil
+}
+
+func (o *CreateSourceSecretOptions) MakeSourceSecret() (*api.Secret, error) {
+	secret := &api.Secret{}
+	secret.Namespace = o.SecretNamespace
+	secret.Name = o.SecretName
+	secret.Type = buildapi.SecretTypeSource
+	secret.Data = map[string][]byte{}
+
+	if len(o.Username) != 0 {
+		secret.Data[buildapi.SourceUsername] = []byte(o.Username)
+	}
+
+	if len(o.Password) != 0 {
+		secret.Data[buildapi.SourcePassword] = []byte(o.Password)
+	}
+
+	if len(o.CertificatePath) != 0 {
+		caContent, err := ioutil.ReadFile(o.CertificatePath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourceCertificate] = caContent
+	}
+
+	if len(o.GitConfigPath) != 0 {
+		gitConfig, err := ioutil.ReadFile(o.GitConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourceGitconfig] = gitConfig
+	}
+
+	if len(o.PrivateKeyPath) != 0 {
+		privateKeyContent, err := ioutil.ReadFile(o.PrivateKeyPath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourcePrivateKey] = privateKeyContent
+	}
+
+	return secret, nil
+}
+
+func (o *CreateSourceSecretOptions) Complete(f *kcmdutil.Factory, args []string) error {
+	if len(args) != 1 {
+		return errors.New("must have exactly one argument: secret name")
+	}
+	o.SecretName = args[0]
+
+	if len(o.Username) != 0 || len(o.Password) != 0 || len(o.CertificatePath) != 0 {
+		if len(o.PrivateKeyPath) != 0 {
+			return errors.New(optionGroupExclusivityError)
+		}
+	}
+
+	if o.PromptForPassword {
+		if len(o.Password) != 0 {
+			return errors.New("must provide either --prompt or --password flag")
+		}
+		if cmdutil.IsTerminal(o.Reader) {
+			o.Password = cmdutil.PromptForPasswordString(o.Reader, o.Out, "Password: ")
+			if len(o.Password) == 0 {
+				return errors.New("password must be provided")
+			}
+		} else {
+			return errors.New("provided reader is not a terminal")
+		}
+	}
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+	o.SecretNamespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	o.SecretsInterface = client.Secrets(o.SecretNamespace)
+
+	return nil
+}
+
+func (o CreateSourceSecretOptions) Validate() error {
+	switch {
```

Comment:
Well this switch does the same trick, but I dont have any strong opinion about not rewriting it to ifs. Do you ? :)